### PR TITLE
MI-6 upload validation

### PR DESCRIPTION
**Pull request by JanionNevill**
To  test the functionality, try uploading any of the files in the `tests/testthat/invalid_data` folder. Further tasks could be to implement validation on the data itself, such as checking that `N` columns contain only integers etc.

This PR will be merged into `feature-meta-regression` after review.

### CODE REVIEW

File: R/DataValidation.R [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+#' Validate wide columns that all columns are numbered sequentially from 1 to the same number across all numbered columns.
+#'
+#' @param uploaded_data Data frame to validate.
+#' @param required_columns Data frame representing required numbered columns. Columns in this data frame are:
+#' - pattern = A regular expression defining the generic column title for wide or long format.
+#' - number_group = Regular expression replacement to extract the number from a column.
+#'
+#' @return TRUE if the columns are correctly numbered, else FALSE.
+.ValidateMatchingWideColumns <- function(uploaded_data, required_columns) {
```
Comment (tommorris168) [inline R/DataValidation.R:83]: Could you add a comment at the top of each block of code in this function to say what it does.

File: tests/testthat/test-DataValidation.R [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+test_that("ValidateUploadedData() identifies valid binary wide data", {
+  data <- CleanData(read.csv("Binary_wide.csv"))
+  
+  validation_result <- ValidateUploadedData(data, "Binary")
+  
+  expect_true(validation_result$valid)
+  expect_equal(validation_result$message, "Data is valid")
+})
+
+test_that("ValidateUploadedData() identifies invalid coninuous long data, missing colunm 'N'", {
```
Comment (tommorris168) [inline tests/testthat/test-DataValidation.R:38]: coninuous (also in several other places)

File: tests/testthat/test-DataValidation.R [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+test_that("ValidateUploadedData() identifies invalid coninuous wide data, with inconsistently numbered colunms 'N.*'", {
+  data <- CleanData(read.csv("invalid_data/continuous-misnumbered-wide.csv"))
+  
+  validation_result <- ValidateUploadedData(data, "Continuous")
+  
+  expect_false(validation_result$valid)
+  expect_equal(validation_result$message, "For wide format data, numbered columns (T, N, Mean, SD) must all have matching sequential indices, starting from 1")
+})
+
+test_that("ValidateUploadedData() identifies invalid coninuous long data, missing colunm 'N'", {
```
Comment (tommorris168) [inline tests/testthat/test-DataValidation.R:74]: Should be binary, also in several other places.

**Review by tommorris168 — CHANGES_REQUESTED** [review]
No comments on the code, just comments on the comments.

File: R/DataValidation.R (hunk 1 above)
Comment (JanionNevill) [reply R/DataValidation.R:83]: Done

File: tests/testthat/test-DataValidation.R (hunk 2 above)
Comment (JanionNevill) [reply tests/testthat/test-DataValidation.R:38]: Fixed

File: tests/testthat/test-DataValidation.R (hunk 3 above)
Comment (JanionNevill) [reply tests/testthat/test-DataValidation.R:74]: Fixed

**Review by tommorris168 — APPROVED** [review]
Approved

**Pull request merged**